OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs

```diff
@@ -196,10 +210,50 @@ void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation)
 
 		void IBotTick.BotTick(IBot bot)
 		{
+			// TODO: this causes pathfinding lag when AI's gets blocked in
 			SetRallyPointsForNewProductionBuildings(bot);
 
-			foreach (var b in builders)
-				b.Tick(bot);
+			BuildingsBeingProduced = new Dictionary<string, int>();
+
+			// PERF: We tick only one type of valid queue at a time
+			// if AI gets enough cash, it can fill all of its queues with enough ticks
+			var findQueue = false;
+			for (int i = 0, builderIndex = currentBuilderIndex; i < builders.Length; i++)
+			{
+				if (++builderIndex >= builders.Length)
+					builderIndex = 0;
+
+				if (--builders[builderIndex].WaitTicks < 0)
+					builders[builderIndex].WaitTicks = 0;
```

Comment:
WaitTick is already set in `BaseBuilderQueueManager`, it only needs -- here

Reply:
there's no harm going into negative

Reply:
I think it is better to have this here, it will be clearer when debug those lines.

Reply:
we can always see `WaitTicks=0` instead of going nagative and becoming a small distraction.

Reply:
Changed now.